### PR TITLE
HPCC-27610 Update Container Placements Documentation

### PR DESCRIPTION
**Pull request by g-pan**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Successful unit testing: http://10.224.20.18/view/Docs-gp/job/DocBuild-GPRepo3/

### CODE REVIEW

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 1]
```diff
@@ -1072,212 +1072,602 @@ thor:
...
+    container in the created pod. Feasible nodes are then scored to find the
+    best candidate for the pod placement. The scheduler also takes into
+    account other factors such as pod affinity and anti-affinity, taints and
+    tolerations, pod topology spread constraints, and the node selector
+    labels. The scheduler can be configured to use these different algorithms
+    and policies to optimize the pod placement according to your cluster’s
+    needs.</para>
+
+    <para>You can deploy these values either using the values.yaml file or you
+    can place into an file and have Kubernetes instead read the values from
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1217]: a file, not an file

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 2]
```diff
@@ -1072,212 +1072,602 @@ thor:
...
+    the supplied file. See the above section <emphasis>Customization
+    Techniques</emphasis> for more information about customizing your
+    deployment.</para>
+
+    <sect2 id="YAML_FileStruct_Placement">
+      <title>Placements</title>
+
+      <para>The Placements is a term used by HPCC Systems, which Kubernetes
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1225]: The Placements term is not our term, it is a standard to Kubernetes. Please reword

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 3]
```diff
@@ -1072,212 +1072,602 @@ thor:
...
-          <para>The pods: [list] can contain a variety of items.</para>
+      <sect3 id="PlacementScope">
+        <title>Placement Scope</title>
+
+        <para>Use pod patterns to apply the scope for the placements.</para>
 
-          <orderedlist>
+        <para>The pods: [list] item can contain one of the following:</para>
+
+        <para><orderedlist>
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1263]: This should not be a numbered list. A table would be a better representation.

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 4]
```diff
@@ -1072,212 +1072,602 @@ thor:
...
-        <title>Environment Values</title>
+    <sect2 id="S2NodeSelection">
+      <title>Node Selection</title>
 
-        <para>You can define environment variables in a YAML file. The
-        environment values are defined under the
-        <emphasis>global.env</emphasis> portion of the provided HPCC Systems
-        values.yaml file. These values are specified as a list of name value
-        pairs as illustrated below.</para>
+      <para>In a Kubernetes container environment there are several ways to
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1324]: needs a comma :
In a Kubernetes container environment, there...

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 5]
```diff
@@ -1072,212 +1072,602 @@ thor:
...
+        list. This example schedules "all" HPCC components to use the node
+        pool with the label group: "hpcc".</para>
+
+        <para><programlisting> placements:
+   - pods: ["all"]
+     placement:
+       nodeSelector:
+         group: "hpcc"</programlisting></para>
+
+        <para>Note:the label: group:hpcc matches the node pool label:
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1396]: Note: s/b bold, then a space and capitalize the sentence. also no need for the quotes around hpcc. 

**Note:** The label: group:hpcc matches the node pool label:hpcc.

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 6]
```diff
@@ -1072,212 +1072,602 @@ thor:
...
 
-                  <row>
-                    <entry>jvmlibpath</entry>
+      <sect3 id="CV-TAINTS_TOLERATIONS">
+        <title>Taints and Tolerations</title>
 
-                    <entry>JAVA_LIBRARY_PATH</entry>
-                  </row>
+        <para>Taints and Tolerations are types of Kubernetes node constraints
+        also referred to by Node Affinity. Only one "affinity" can be applied
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1421]: inconsistent capitalization of Node Affinity. Previously, it was node Affinity.

Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1421]: no need for quotes around affinity (2 places)

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 7]
```diff
@@ -1072,212 +1072,602 @@ thor:
...
-  numWorkers: 2
-  maxJobs: 4
-  maxGraphs: 2
-  indexBuildPlane: myindexplane</programlisting>
+        <para>Using the "topologySpreadConstraints" example, there are two
+        node pools which have "hpcc=nodepool1" and "hpcc=nodepool2"
+        respectively. The Roxie pods will be evenly scheduled on the two node
+        pools.</para>
+
+        <para>After deployment you can verify by issuing the following:</para>
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1527]: After deployment you can verify by issuing the following command:

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 8]
```diff
@@ -1072,212 +1072,602 @@ thor:
...
+          can receive a pod by matching labels of those nodes. Node affinity
+          and anti-affinity can only be used to set positive affinities that
+          attract pods to the node. These are used to constrain the nodes that
+          can receive a pod by matching labels to those nodes. Node affinity
+          and anti-affinity can only be used to set positive affinities that
+          attract pods to the node.</para>
+
+          <para>There is no schema check for the content of affinity. Only one
+          affinity can be applied to a pod or job. If a pod/job matches
+          multiple placement 'pods' lists, then only the last affinity
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1573]: no need for the quotes around 'pods'

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 9]
```diff
@@ -1072,212 +1072,602 @@ thor:
...
+
+        <para>The <emphasis role="bold">schedulerName</emphasis> field
+        specifies the name of the scheduler that is responsible for scheduling
+        a pod or a task. In Kubernetes, you can configure multiple schedulers
+        with different names and profiles to run simultaneously in the
+        cluster.</para>
+
+        <para>Only one "schedulerName" can be applied to any pod/job.</para>
+
+        <para>A SchedulerName example:</para>
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1655]: Inconsistent capitalization of schedulerName

**Review by JamesDeFabia — CHANGES_REQUESTED** [review]
A few comments inline

**Review by xwang2713 — APPROVED** [review]
Looks good for me. Just Remove "Examples:" before "Taints and Tolerations" on p20

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 10]
```diff
@@ -1072,212 +1072,621 @@ thor:
...
+    <sect2 id="YAML_FileStruct_Placement">
+      <title>Placements</title>
+
+      <para>The Placements is a term used by HPCC Systems, which Kubernetes
+      refers to as the scheduler or scheduling/assigning. In order to avoid
+      confusion within the HPCC Systems and ECL specific scheduler terms,
+      refer to Kubernetes scheduling as placements. Placements is a value in
+      an HPCC Systems configuration which is at a level above items, such as
+      the nodeSelector, Toleration, Affinity and Anti-Affinity, and
+      TopologySpreadConstratints.</para>
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1231]: DId you mean TopologySpreadConstraints?

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 11]
```diff
@@ -1072,212 +1072,621 @@ thor:
...
+
+      <para>The placement is responsible for finding the best node for a pod.
+      Most often placement are handled automatically by Kubernetes. You can
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1234]: S/V agreement: S/B Most often placements are handled

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 12]
```diff
@@ -1072,212 +1072,621 @@ thor:
...
+    <para>You can deploy these values either using the values.yaml file or you
+    can place into a file and have Kubernetes instead read the values from the
+    supplied file. See the above section <emphasis>Customization
+    Techniques</emphasis> for more information about customizing your
+    deployment.</para>
+
+    <sect2 id="YAML_FileStruct_Placement">
+      <title>Placements</title>
+
+      <para>The Placements is a term used by HPCC Systems, which Kubernetes
```
Comment (JamesDeFabia) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1225]: remove word "The"

**Review by JamesDeFabia — CHANGES_REQUESTED** [review]
Couple of minor comments inline

**Review by JamesDeFabia — APPROVED** [review]
Looks good from my POV

**Review by JamesDeFabia — APPROVED** [review]
Looks good from my POV

**Review by JamesDeFabia — APPROVED** [review]
Looks OK from my POV

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 13]
```diff
@@ -1072,212 +1072,621 @@ thor:
...
+                targets with each having a unique name such as "roxie",
+                "roxie2", "roxie-web" etc.</entry>
+              </row>
+
+              <row>
+                <entry>Pod: &lt;name&gt;</entry>
+
+                <entry>This is the "Deployment" metadata name from the name of
+                the array item of a type. For example, "eclwatch-", "mydali-",
+                "thor-thoragent" This can be a regular expression since
```
Comment (jakesmith) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1296]: @xwang2713 - are regular expressions supported or just wildcards?

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 14]
```diff
@@ -1072,212 +1072,621 @@ thor:
...
+                eclwatch-7f4dd4dd44cb-c0w3x.</entry>
+              </row>
+
+              <row>
+                <entry>Job name:</entry>
+
+                <entry>The job name is typically a regular expression as well,
+                since the job name is generated dynamically. For example, a
+                compile job compile-54eB67e567e, could use "compile-" or
+                "compile-*" or the exact match "^compile-.$"</entry>
```
Comment (jakesmith) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1308]: @xwang2713 - I want to clarify if this supports regular expressions or just wildcards, "compile-*" is not an example of a regular expression (well it is, but would match 'compile----' , or 'compile', not 'compile-blah')

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 15]
```diff
@@ -1072,212 +1072,621 @@ thor:
...
+        configuration, they will be processed in the following order: "all",
+        "type", "target", and then "pod"/"job".</para>
+
+        <sect4>
+          <title>Mixed combinations</title>
+
+          <para>NodeSelector, taints and tolerations, and other values can all
+          be placed on the same pods: [list] both per zone and per node on
+          Azure <programlisting>placements:
+- pods: ["eclwatch","roxie-workunit","^compile-.*$","mydali"]
```
Comment (jakesmith) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1331]: but 'compile-.*$' is a regex, so is it that it does support regex's, but the example 'compile-*' is wrong?

**Review by jakesmith — COMMENTED** [review]
@g-pan - please see question re. regex/pattern match.

**Comment by ghalliday** [comment]
Assigned to @xwang2713 to answer Jake's questions.

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml (hunk 13 above)
Comment (xwang2713) [reply docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1296]: The regular expressions are preferred. But I think  substrings are also OK .  Helm docs (mustRegexMatch):
regexMatch, mustRegexMatch. Returns true if the input string contains any match of the regular expression.

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml (hunk 14 above)
Comment (xwang2713) [reply docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1308]: Either "compile-.*" or "compile-".  "compile-*" is not valid regular expression or substring

Comment (jakesmith) [reply docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1308]: Ok, then line 1308 needs changing to "compile-.\*"
@g-pan

**Review by jakesmith — COMMENTED** [review]
@g-pan - see follow-up comment.

**Comment by g-pan** [comment]
added the last change and squashed (since it was minor) should be ready to merge now.

File: docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml [hunk 16]
```diff
@@ -1072,212 +1072,621 @@ thor:
...
+                eclwatch-7f4dd4dd44cb-c0w3x.</entry>
+              </row>
+
+              <row>
+                <entry>Job name:</entry>
+
+                <entry>The job name is typically a regular expression as well,
+                since the job name is generated dynamically. For example, a
+                compile job compile-54eB67e567e, could use "compile-" or
+                "compile-.*" or "^compile-.$"</entry>
```
Comment (jakesmith) [inline docs/EN_US/ContainerizedHPCC/ContainerizedMods/ConfigureValues.xml:1308]: @g-pan - still not quite right, missing * before $, see previous comment

**Review by jakesmith — COMMENTED** [review]
@g-pan - back to you.

**Comment by g-pan** [comment]
> @g-pan - back to you.

good catch!

**Review by jakesmith — APPROVED** [review]
@g-pan - looks good now.

**Pull request merged**